source/module_relax/relax_old/test/bfgs_basic_test.cpp

```diff
@@ -0,0 +1,376 @@
+#include "module_relax/relax_old/ions_move_basic.h"
+#define private public
+#define protected public
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+#include "module_relax/relax_old/bfgs_basic.h"
+/************************************************
+ *  unit tests of class BFGS_Basic
+ ***********************************************/
+
+/**
+ * - Tested Functions:
+ *   - BFGS_Basic::allocate_basic()
+ *   - BFGS_Basic::new_step()
+ *   - BFGS_Basic::reset_hessian()
+ *   - BFGS_Basic::save_bfgs()
+ *   - BFGS_Basic::check_move()
+ *   - BFGS_Basic::update_inverse_hessian()
+ *   - BFGS_Basic::check_wolfe_conditions()
+ *   - BFGS_Basic::compute_trust_radius()
+ */
+
+int Ions_Move_Basic::dim = 0;
+bool Ions_Move_Basic::converged = false;
+double Ions_Move_Basic::largest_grad = 0.0;
+int Ions_Move_Basic::update_iter = 0;
+int Ions_Move_Basic::istep = 0;
+double Ions_Move_Basic::ediff = 0.0;
+double Ions_Move_Basic::etot = 0.0;
+double Ions_Move_Basic::etot_p = 0.0;
+double Ions_Move_Basic::trust_radius = 0.0;
+double Ions_Move_Basic::trust_radius_old = 0.0;
+double Ions_Move_Basic::relax_bfgs_rmax = -1.0;
+double Ions_Move_Basic::relax_bfgs_rmin = -1.0;
+double Ions_Move_Basic::relax_bfgs_init = -1.0;
+double Ions_Move_Basic::best_xxx = 1.0;
+int Ions_Move_Basic::out_stru = 0;
+
+double Ions_Move_Basic::dot_func(const double *a, const double *b, const int &dim_in)
+{
+    double result = 0.0;
+    for (int i = 0; i < dim_in; i++)
+    {
+        result += a[i] * b[i];
+    }
+    return result;
+}
+
+class BFGS_Basic_Test : public ::testing::Test
+{
+  protected:
+    void SetUp() override
+    {
+        // Initialize variables before each test
+    }
+
+    void TearDown() override
+    {
+        // Clean up after each test
+    }
+
+    BFGS_Basic bfgs;
+};
+
+// Test whether the allocate_basic() function can correctly allocate memory space
+TEST_F(BFGS_Basic_Test, TestAllocate)
+{
+    Ions_Move_Basic::dim = 4;
+    bfgs.allocate_basic();
+
+    // Check if allocated arrays are not empty
+    EXPECT_NE(nullptr, bfgs.pos);
+    EXPECT_NE(nullptr, bfgs.pos_p);
+    EXPECT_NE(nullptr, bfgs.grad);
+    EXPECT_NE(nullptr, bfgs.grad_p);
+    EXPECT_NE(nullptr, bfgs.move);
+    EXPECT_NE(nullptr, bfgs.move_p);
+    EXPECT_NE(nullptr, bfgs.inv_hess.c);
+}
+
+// Test if a dimension less than or equal to 0 results in an assertion error
+TEST_F(BFGS_Basic_Test, TestAllocateWithZeroDimension)
+{
+    Ions_Move_Basic::dim = 0;
+    ASSERT_DEATH(bfgs.allocate_basic(), "");
+}
+
+// Test function update_inverse_hessian() assert death
+TEST_F(BFGS_Basic_Test, Update_Inverse_Hessian_Death)
```

Comment:
Googletest suggests against using underscores in naming tests (https://google.github.io/googletest/primer.html). I don't know to exact reason, but we should better do as said to avoid possible errors.